[Docs] cts documentation (#3698)
The-OpenROAD-Project/OpenROAD: src/cts/README.md

`configure_cts_characterization` command. Use `set_wire_rc` command to 
set clock routing layer.

## Commands

```{note}
- Parameters in square brackets `[-param param]` are optional.
- Parameters without square brackets `-param2 param2` are required.
```

### Configure CTS Characterization

```tcl
configure_cts_characterization 
    [-max_slew max_slew]
    [-max_cap max_cap]
    [-slew_steps slew_steps]
    [-cap_steps cap_steps]
```

#### Options

| Switch Name | Description | 
| ----- | ----- |
| `-max_slew` | Max slew value (in seconds) that the characterization will test. If this parameter is omitted, the code would use max slew value for specified buffer in `buf_list` from liberty file. |
| `-max_cap` | Max capacitance value (in farad) that the characterization will test. If this parameter is omitted, the code would use max cap value for specified buffer in `buf_list` from liberty file. |
| `-slew_steps` | Number of steps that `max_slew` will be divided into for characterization. (default 12) |
| `-cap_steps` | Number of steps that `max_cap` will be divided into for characterization. (default 34) |

### Clock Tree Synthesis

```tcl

> clock_tree_synthesis 

Reviewer: Missing:
```
[-num_static_layers] \
[-sink_clustering_buffer] \
```

Author: I need some clarification on this, as I am not sure on the purpose on these switches.

Can you help me ping the tool maintainer?

Reviewer: He is on vacation this week; we can keep a generic description and improve it later. This is good enough for now, thanks.

    -buf_list <list_of_buffers>
    [-root_buf root_buf]
    [-wire_unit wire_unit]
    [-clk_nets <list_of_clk_nets>]
    [-distance_between_buffers]
    [-branching_point_buffers_distance]
    [-clustering_exponent]
    [-clustering_unbalance_ratio]
    [-sink_clustering_enable]
    [-sink_clustering_size cluster_size]
    [-sink_clustering_max_diameter max_diameter]
```

#### Options

| Switch Name | Description |
| ----- | ----- |
| `-buf_list` | Tcl list of master cells (buffers) that will be considered when making the wire segments (e.g. `{BUFXX, BUFYY}`). |
| `-root_buffer` | The master cell of the buffer that serves as root for the clock tree. If this parameter is omitted, the first master cell from `-buf_list` is taken. |
| `-wire_unit` | Minimum unit distance between buffers for a specific wire. If this parameter is omitted, the code gets the value from ten times the height of `-root_buffer`. |

> | `-clk_nets` | String containing the names of the clock roots. If this parameter is omitted, TritonCTS looks for the clock roots automatically. |

Reviewer: ```suggestion
| `-clk_nets` | String containing the names of the clock roots. If this parameter is omitted, `cts` automatically looks for the clock roots. |
```

> | `-distance_between_buffers` | Distance (in microns) between buffers that TritonCTS should use when creating the tree. When using this parameter, the clock tree algorithm is simplified, and only uses a fraction of the segments from the LUT. |

Reviewer: ```suggestion
| `-distance_between_buffers` | Distance (in microns) between buffers that `cts` should use when creating the tree. When using this parameter, the clock tree algorithm is simplified and only uses a fraction of the segments from the LUT. |
```

| `-branching_point_buffers_distance` | Distance (in microns) that a branch has to have in order for a buffer to be inserted on a branch end-point. This requires the `-distance_between_buffers` value to be set. |
| `-clustering_exponent` | Value that determines the power used on the difference between sink and means on the CKMeans clustering algorithm. (default 4, integer). |

> | `-clustering_unbalance_ratio` | Value that determines the maximum capacity of each cluster during CKMeans. A value of 50% means that each cluster will have exactly half of all sinks for a specific region (half for each branch). (default 0.6, float). |

Reviewer: ```suggestion
| `-clustering_unbalance_ratio` | Value determines each cluster's maximum capacity during CKMeans. A value of `0.5` (i.e., 50%) means that each cluster will have exactly half of all sinks for a specific region (half for each branch). (default 0.6, float). |
```

| `-clustering_exponent` | Value that determines the power used on the difference between sink and means on the CKMeans clustering algorithm. (default 4, integer). |
| `-clustering_unbalance_ratio` | Value that determines the maximum capacity of each cluster during CKMeans. A value of 50% means that each cluster will have exactly half of all sinks for a specific region (half for each branch). (default 0.6, float). |
| `-sink_clustering_enable` | Enables pre-clustering of sinks to create one level of sub-tree before building H-tree. Each cluster is driven by buffer which becomes end point of H-tree structure. |
| `-sink_clustering_size` | Specifies the maximum number of sinks per cluster. (default 20, integer). |
| `sink_clustering_max_diameter` | Specifies maximum diameter (in microns) of sink cluster (default 50, integer). |
| `-clk_nets` | String containing the names of the clock roots. If this parameter is omitted, TritonCTS looks for the clock roots automatically. |

### Report CTS

Another command available from TritonCTS is `report_cts`. It is used to

> extract metrics after a successful `clock_tree_synthesis` run. These

Reviewer: As discussed, the same of the original work should be mentioned in the first paragraph of the readme. In all other places we should refer to the OR module name, in this case `cts`.

> | `-out_file` | The file containing the TritonCTS reports. If this parameter is omitted, the metrics are streamed to `stdout` and not saved. |

Reviewer: ```suggestion
| `-out_file` | The file to save `cts` reports. If this parameter is omitted, the report is streamed to `stdout` and not saved. |
```

# Clock Tree Synthesis

> The clock tree synthesis module in OpenROAD (`cts`) is based on TritonCTS 
> 2.0. It is available from the `clock_tree_synthesis` command.  
> TritonCTS 2.0 performs on-the-fly characterization. Thus there is
> no need to generate characterization data. On-the-fly characterization feature
> can be optionally controlled by parameters specified to 
> `configure_cts_characterization` command. Use `set_wire_rc` command to 
> set clock routing layer.

Reviewer: ```suggestion
The clock tree synthesis module in OpenROAD (`cts`) is based on TritonCTS
2.0. It is available from the `clock_tree_synthesis` command. TritonCTS 2.0
performs on-the-fly characterization. Thus, there is no need to generate
characterization data. The on-the-fly characterization feature can be optionally
controlled by parameters specified by the `configure_cts_characterization`
command. Use `set_wire_rc` command to set the clock routing layer.
```

| `-sink_clustering_size` | Specifies the maximum number of sinks per cluster. The default value is `20`, and the allowed values are integers `[0, MAX_INT]`. |
| `sink_clustering_max_diameter` | Specifies maximum diameter (in microns) of sink cluster. The default value is `50`, and the allowed values are integers `[0, MAX_INT]`. |
| `-clk_nets` | String containing the names of the clock roots. If this parameter is omitted, `cts` looks for the clock roots automatically. |
| `-num_static_layers` | Set the number of static layers. The default value is `0`, and the allowed values are integers `[0, MAX_INT]`. |
| `-sink_clustering_buffer` | Set the sink clustering buffer(s) to be used. |

### Report CTS

Another command available from `cts` is `report_cts`. It is used to
extract metrics after a successful `clock_tree_synthesis` run. These are:
 
- Number of Clock Roots
- Number of Buffers Inserted
- Number of Clock Subnets
- Number of Sinks.  

```tcl
report_cts 
    [-out_file file]
```

#### Options

| Switch Name | Description |
| ----- | ----- |
| `-out_file` | The file to save `cts` reports. If this parameter is omitted, the report is streamed to `stdout` and not saved. |

> ### Useful developer functions

Reviewer: ```suggestion
### Useful Developer Commands
```

If you are a developer, you might find these useful. More details can be found in the [source file](./src/TritonCTS.cpp) or the [swig file](./src/TritonCTS.i).

```
clock_tree_synthesis_debug -plot # choose to plot the CTS to GUI
```

## Example scripts

```tcl
clock_tree_synthesis -root_buf "BUF_X4" \
                     -buf_list "BUF_X4" \
                     -wire_unit 20
report_cts "file.txt"
```

## Regression tests

> There are a set of regression tests in `/test`. For more information, refer to this [section](../../README.md#regression-tests).

Reviewer: ```suggestion
There are a set of regression tests in `./test`. For more information, refer to this [section](../../README.md#regression-tests).
```
